### ISSUE TITLE
Windows not automatically connecting to Wifi when booting out of shutdown or sleep

### ISSUE THREAD

**Issue by narration-sd** [mullvad/mullvadvpn-app#2352]
@faern Linus, I've written the project, and you yourself, as  a software professional a number of times before, but this business is serious, beyond attitudes and push-offs to 'support'.

Of course I brought the problem to Support at the very first. They were, unsurprisingly, unable to help. And now I have the latest Betas, which work fine on both Android phone and Windows.  With one exception:  

Something like 30-40% of the time, Mullvad is 'Blocked' and never opens its connections, on boot of a Windows 10 machine.

This machine is powerful, well-provisioned, and very, very up-to-date, as I do programming on it (EU Open Source projects via TU Delft). Its only 'difference', which may well affect these bad results with Mullvad, may be that it has a number of startup programs, due to the tools I use.

What is clear in happening is that Mullvad's app tries _once_, to see if it has internet to work with. If not, it _never_ retries. Thus, it  never connects, until I manually use the Disconnect button, wait until internet is clearly established from wifi, and then manually click Secure Connection, at which point it always does operate properly.

Any sensible programming in the environment of networks knows they may not be reliable -- can go up and down, can have limited throughput, etc..  Thus every protocol has retries -- TCP/IP as a prime and now primitive example.

Mullvad performs nicely in low-bandwidth scenarios, but it does not perform at all in the initially blocked internet situation. Which is going to occur on any computer if Mullvad isn't somehow automatically sequenced to try only after internet is established. Which is not an available mechanism from outside Mullvad, on probably any computer, but surely Windows.

There are so many ways to discern if a port is open, watch it, etc., and even more means to code a dependable try-and-try-again-if-you-don't-succeed mechanism. I've proposed one, that you simply try every ten seconds (or of course settable interval) until Mullvad has the internet, and so can form the tunnel. 

Yes, there are considerations as far as how long to try, etc.. Soluble by putting up an alert when trials run out after a truly adequate ttime, say two minutes, but again, should be in the advanced Settings to be able to tune it.

I'm very confident you understand all the things mentioned and more besides. I'm also confident we both would react with the same sense of urgency, for others as well as ourselves, if we saw this missing aspect of Mullvad's system design causing a problem that interrupted our own work.

Thank you for a bit of Christmas spirit, in bringing the simple close to this problem now. I and I suspect silent others will appreciate  it, and it will certainly improve Mullvad's profile as far as reviews, your business partners, and so forth. 

I am in fact surprised there haven't been more of those reviews and very positive, as Mullvad performs very nicely, once it actually starts. One might even think reviewers are giving you a kindly 'pass', not wanting to say anything until they feel that the software is fully complete, and ready for wider public attention on its release.

Kind regards, and a nice amount of that Christmas spirit to you and team, in the time of Santa Lucia,
Clive

**Comment by faern** [comment]
The Windows computer where you see this issue, does it have internet connectivity over Wifi or cabled network? We have a few more reports about people seeing similar things on Wifi. So I want to know if your setup is the same or if the same thing can happen on ethernet via cable in Windows. When your computer boots and you end up in this blocked state you are describing, can you see if Windows has established network connectivity or not? The normal Windows network tray icon should show the status.

Do you have the "Local network sharing" setting enabled or disabled (It's under Settings -> Preferences)? If you have it disabled (the default), would you mind trying to enable it and see if that solves the issue?

The Mullvad VPN app tries over and over to establish a connection. But not if it detects that the computer is offline. Then it sits and wait for the computer to come online. So what I suspect is happening here is that our app somehow blocks the Wifi from connecting. When you disable Mullvad the Wifi can connect and after that Mullvad can connect fine. That's why you interpret Mullvad as not even trying over and over. Because our app can see that no network adapter is connected, so there is no idea in even trying, all it can do is wait for a network adapter to come active.

Please answer the questions in a succinct manner please. Providing lengthy explanations of how network software is supposed to work is not going to be an efficient way towards a solution, thank you.

**Comment by narration-sd** [comment]
Hi Linus, here are answers:
- I'm accessing via wifi
- in both present location and where earlier reporting occurred, it's over a cable internet.  This one is faster - 50Mbps vs. 10Mbps
- indeed, the basis wifi connection has not occurred yet, when Mullvad gets to run first, and finds its conditions for 'Block'

I suspect you're right about discovering the network adapter being the root of the problem. t

- One reason to think so is that if wifi service is interrupted later, then comes back on, Mullvad does properly notice and reconnect.
- However, another factor I notice is that if I _disable_ the Wifi 'adapter' (driver) in Win10 Network Connections, then bring it back after a pause, Mullvad also properly notices and reconnects.
- so, there might be a clue how to notice the original driver initiation. Though, the problem might be something like that a list of drivers isn't even available, for the instances where Mullvad isn't connecting on boot.

Hope these help, and thanks, Linus

**Comment by narration-sd** [comment]
^^edited a bit in GitHub, added information, in case you're reading only over email at first...

**Comment by faern** [comment]
Thanks for clarifying this problem is over Wifi. Re-trying is not part of the problem I think. I think the entire problem stems from Mullvad somehow blocking the Wifi from establishing a connection. And obviously that will leave everything without internet. Mullvad is not part of anything related to wifi or drivers. But something we do seem to prevent Windows from establishing a wifi connection I think. We have logged it as an issue and will investigate, thank you.

**Comment by narration-sd** [comment]
Thank you, Linus, and that's an interesting view -- that Mullvad is actually blocking the Wifi driver from making the connection.

Intrigued by what you may find out, and greetings in the season, for all of you and families there.
Clive

**Comment by narration-sd** [comment]
@faern  Linus, thought I noticed something, so I made a test, and I think it may validate your idea that the problem is the Mullvad app/driver somehow preventing wifi from actually connecting.

- I rebooted twice, to get the state of Mullvad 'blocked'
- noted, that the indicator for internet/wifi appeared pre-connection state
- indeed, clicked on it (lower right mini-icons), and saw my wifi with nearby others _available_ for connection, but not following its checkmark to auto-connect
- I clicked to connect to my wifi, and it immediately did connect
- as soon as the wifi was connected, Mullvad itself auto-created the tunnel to the VPN site saved (Los Angeles, a particular node)

So it looks like you are rather likely to be on the correct track here.

Holiday thoughts, and this can wait of course :)
Clive

**Comment by windowsdeveloperwannabe** [comment]
I've not had any trouble on Desktop as it uses ethernet, but waking up a laptop on wifi and VPN is a classic annoyance. Sometimes after wake Mullvad (like other VPN apps before it) just keeps blocking the Internet and itself until you manually hit disconnect and try again.

**Comment by narration-sd** [comment]
Good morning, after then holidays...

- I notice the title had changed, but this was not in my case about waking from sleep -- it's on full Win10 boot.

- and, I have a further clue.. Over the holiday week, at any hour, the booting connection worked without incident.

  But on this Monday mid-morning, it failed and perhaps a little harder than usual, so that I was able to gather information:
  = Mullvad was'blocked'
  = I was able to try releasing and reconnecting the laptop wifi, via the menu from the usual small icon on the right of the Task bar
  = the fresh connection occurred, but after a slight delay it informed that it was connected, but with 'no internet' 
  = trying again didn't improve on this
  = however, when I told Mullvad to Disconnect, immediately the intertnet reported as connected. I believe I had disconnected before disabling Mullvad, so that this indication occurred on the fresh wifi connect happening.
  = as ever, once fully connected, Mullvad worked fine when told to connect for itself

You can see if this sequence offers anything fresh. I am probably influenced by reading another issue here, to wonder if the problem may have something to do with Mullvad re-routing or otherwise blocking DNS. How, we can question, does Win10 become aware that it has indeed internet, besides a wifi connection?  One may think it would take a DNS call to find out; hence the possible hint.

Best fortune...

**Comment by dlon** [comment]
@mullvadanonuser 
> [2020-12-29 12:26:37.660][Connectivity monitor][INFO] Connectivity changed. Machine is: ONLINE
> [2020-12-29 12:26:38.113][talpid_core::offline::imp][DEBUG] Machine is returning from sleep mode
> [2020-12-29 12:26:43.139][talpid_core::offline::imp][DEBUG] TAP is presumed to have been re-initialized
> [2020-12-29 12:26:43.139][mullvad_daemon::relays][DEBUG] Selecting among 16 relays with combined weight 3000
> [2020-12-29 12:26:48.405][talpid_core::tunnel_state_machine::connecting_state][ERROR] Error: Failed to start tunnel
> Caused by: Failed while listening for events from the Wireguard tunnel
> Caused by: Tunnel failed
> Caused by: Failed to start wireguard tunnel

Interesting. Looks like the offline detection is working fine here (unless it incorrectly assumes that the device is online). What fails is setting up the WireGuard tunnel. Would you mind sharing `wireguard.log`, if it contains anything useful?

**Comment by narration-sd** [comment]
sure, sharing what logs I could find, which were under c:\Users\username\AppData\Local\Mullvad, attached.

[mullvad-logs.zip](https://github.com/mullvad/mullvadvpn-app/files/5783870/mullvad-logs.zip)

I wasn't able to locate any 'wireguard.log' in any of the normal places.

From logs in the zip, you can see this morning's login which failed, and only was allowed after I told Mullvad not to disconnect (from blocked), but rather to change Location -- then re-connected wifi, which then and only then said it had internet access. Mullvad immediately built the tunnel on that connection to the changed location. I changed it back, thus the reconnect in logs.

The .old log versions show a proper immediate connect and tunnel from a login at a previous time.

**Comment by dlon** [comment]
@narration-sd I don't know if your problems are identical, but the relevant logs can be found in `C:\ProgramData\Mullvad VPN`.

**Comment by narration-sd** [comment]
Ok, this may be interesting. 

I had trouble of course getting the start to go wrongly, but when I did, something unusual happened.  Mullvad and thus internet connectivity blocked, but then as I've never seen it do before, some portion of a minute later, it unblocked and all connected.

Thus, these logs show the original blockage, and this recovery.

[Mullvad-lock-unlock.zip](https://github.com/mullvad/mullvadvpn-app/files/5788969/Mullvad-lock-unlock.zip)

I've included two files, because wireguard.log doesn't look interesting for this problem -- it just shows what happens when an actual proper connection occurs. But the daemon.log gives at least some details of the block and then release, so that may help.

Trying to parse meanings of sequence items, it occurred to me that maybe the problem of blocking occurs if Mullvad runs early and rapidly in the startup sequence of many apps and threads. 

If Mullvad's blocking action stated in daemon.log actually affects the wifi driver's ability to connect, that could explain much.  

Thus, a race, which is what the problem looks like in observing it, again on hardware with enough power to run multiple processes and threads rapidly - i7-4700MQ, good SSD, etc..

**Comment by narration-sd** [comment]
And here's a pair of logs for another instance of the problem - once again, self-correcting after a while, which seems new behavior, though I'm still on the 2020.8-beta2 since it was announced 11 December.

[Mullvad-2-lock-unlock.zip](https://github.com/mullvad/mullvadvpn-app/files/5789306/Mullvad-2-lock-unlock.zip)

There's somewhat more action happening up front on the daemon.log here; not sure if that's part of what's relevant.

**Comment by narration-sd** [comment]
@faern  Linus, I think I have something else to report about this problem, and it might help if there's been difficulty duplicating it.

First, as I watch what happens on booting the laptop these days, I think the logs I've sent don't directly show the problem, or at least the second one doesn't. 

What it shows is what I see over and over now, dependably since at least the day I posted that second log
1) the laptop starts, and I log in
2) on early screen, Mullvad posts that it can't operate, suggests connecting to the internet in a small toast
3) soon thereafter, say 5-10 seconds, Mullvad does connect, all by itself, anitd says now that it has
4) all is ok from that point, including automatic reconnection if the wifi goes down and back up, for example

Somehow, then, now there is not the original fully-locked-out-from-boot-up problem.

Why?  One would think maybe Microsoft sent a patch. If so, it wasn't  the second Tuesday for January, because this better behavior began well before that. And usually, patches are not silent, though of course they can and do do as they please.

The other clue in this is that I never saw any toast suggesting wifi connections in the original problem. Never. This message is entirely new - possibly included in your latest beta, but again, I never saw it in the original lockout problem, which continued after installing that beta the day it was released.

I'm a little onplussed, though we know there will be a reason somewhere. Anyway, I'm informing you, and we can watch the situation if you haven't located anything that could have influenced to enable the original problem. 

Thanks for your attention, in any case, and after a while, if no-one else still has the problem, of course this issue can become closed.,

**Comment by narration-sd** [comment]
...just corrected again the title, as my reporting has always been about fresh booting, never about sleep.  I do see that others have had an issue with sleep, which may well be the same underlying problem, of course

**Comment by narration-sd** [comment]
@faern  Finally, Linus, believe I caught a log of the issue happening. It's attached,, with logs and old_logs.

- I booted the machine
- this time,, Mullvad 'stuck' as it had in the original report/s, on a boot.
- it did not recover at all, just left the toast reporting failure
- I opened the Mullvad interface, and cancelled the attempt
- a few moments later, I clicked Secure my Connection
- now, as in the original report, Mullvad connected right away
- an hour or so later, I closed the driver for the wifi, and Mullvad properly indicated
- directly after, I opened the driver to the wifi, it connected, and then Mullvad connect its tunnel smoothly, as it always does

I'm not sure if there are useful clues, looking at the logs, but I hope so...

[Mullvad-fail-on-boot-manual-recovery.zip](https://github.com/mullvad/mullvadvpn-app/files/5891477/Mullvad-fail-on-boot-manual-recovery.zip)

**Comment by narration-sd** [comment]
@faern   Linus, more on this. It looks like the problem for whatever reason is becoming prevalent again, as once again Mullvad blocked on booting the laptop this morning.

This time I've zipped logs while the block is present, and then after I exited and restarted Mullvad, where it happily connected. 

I noticed from email delivery that the actual wifi port became unblocked as soon as Mullvad had been exited, which fits what we seem to know about this to a point, but I didn't think to examine the state of the wifi by other means while Mullvad was blocked -- will do that next time.

Here are the logs. They're named _copy because with Mullvad running ,they were 'in use; so couldn't be zipped. Further, if it's any clue, one of the logs was still 'in use' after I exited Mullvad this time -- likely the daemon.log, imagine, and if any possible clue that daemon might not have exited.

[Mullvad fail step only 29jan2021.zip](https://github.com/mullvad/mullvadvpn-app/files/5895473/Mullvad.fail.step.only.29jan2021.zip)
[Mullvad success step after 29jan2021.zip](https://github.com/mullvad/mullvadvpn-app/files/5895474/Mullvad.success.step.after.29jan2021.zip)

**Comment by narration-sd** [comment]
...and, @faern, even more information, as Mullvad really stuc,k hard earlier today on a boot-up. I couldn't get it to connect by the usual means, shutting down and reopening the wifi connection, , which on other occasions has let Mullvad connect its tunnel. 

- in this case, I had to cancel Mullvad, then ask it again to Secure, which it then did
- attached, are logs before and after the blockage
- I also took ipconfig listings before and after. The only difference, other than temp ipv6 address on the actual wifi (final address was the same), was that once connected, the Mullvad driver showed up:
> Unknown adapter Mullvad:

   Connection-specific DNS Suffix  . :
   Link-local IPv6 Address . . . . . : fe80::34d4:b408:b53b:b7f6%35
   IPv4 Address. . . . . . . . . . . : 10.67.75.99
   Subnet Mask . . . . . . . . . . . : 255.255.255.255
   Default Gateway . . . . . . . . . :

You can see all the action where I tried dropping and reconnecting the wifi itself in the logs. Only in the very last portion for the daemon do you see the tunnel connecting, after I'd canceled and re-Secured Mullvad.

[Mullvad hard blocked.zip](https://github.com/mullvad/mullvadvpn-app/files/5899028/Mullvad.hard.blocked.zip)
[Mullvad recovered from hard block.zip](https://github.com/mullvad/mullvadvpn-app/files/5899029/Mullvad.recovered.from.hard.block.zip)

Good hunting...!

**Comment by dabljues** [comment]
I can confirm that this also happens when using ethernet connection - I have to cancel mullvad trying to connect and connect to a server manually

**Comment by narration-sd** [comment]
This has been fully removed by the excellent work of @faern @dlon and anyone else on the Mullvad team.

See #2436 for details
